File: legacy/cadence_legacy/lib/cadence_web/live/ops_console_live/widgets/alarm_widget.ex
defmodule CadenceWeb.OpsConsoleLive.Widgets.AlarmWidget do
  @moduledoc """
  Phoenix LiveComponent for displaying alarm summaries.

  Shows filtered alarm list with:
  - Severity counts (critical/warning/info)
  - Inline actions (acknowledge, shelve, clear)
  - Configurable filters

  ## Usage

      <.live_component
        module={AlarmWidget}
        id="alarm-widget-1"
        widget_id="widget-789"
        alarms={@active_alarms}
        config={%{
          "severity_filter" => ["critical", "warning"],
          "target_filter" => nil,
          "max_alarms" => 20
        }}
        on_configure={JS.push("open_widget_config")}
        on_acknowledge={JS.push("acknowledge_alarm")}
        on_shelve={JS.push("open_shelve_modal")}
        on_clear={JS.push("clear_alarm")}
      />
  """
  use CadenceWeb, :live_component

  import CadenceWeb.WidgetComponents

  alias Phoenix.LiveView.JS

  @impl true
  def mount(socket) do
    {:ok, assign(socket, acknowledged_ids: MapSet.new())}
  end

  @impl true
  def update(assigns, socket) do
    {:ok, assign(socket, assigns)}
  end

  @impl true
  def render(assigns) do
    config = assigns[:config] || %{}
    alarms = assigns[:alarms] || []
    title = config["title"] || "Alarms"

    # Apply filters
    severity_filter = config["severity_filter"] || []
    target_filter = config["target_filter"]
    max_alarms = config["max_alarms"] || 20
    show_cleared = config["show_cleared"] || false

    filtered_alarms =
      alarms
      |> filter_by_severity(severity_filter)
      |> filter_by_target(target_filter)
      |> filter_cleared(show_cleared)
      |> Enum.take(max_alarms)

    # Calculate counts from unfiltered (but not cleared) alarms
    active_alarms = Enum.reject(alarms, &(&1.status == :cleared))
    counts = calculate_counts(active_alarms)

    has_filters = severity_filter != [] || target_filter != nil

    assigns =
      assigns
      |> assign(:title, title)
      |> assign(:filtered_alarms, filtered_alarms)
      |> assign(:counts, counts)
      |> assign(:has_filters, has_filters)

    ~H"""
    <.widget
      id={@widget_id}
      title={@title}
      on_configure={
        @on_configure &&
          JS.push("open_widget_config", value: widget_config_value(@widget_id, "alarm", @config))
      }
    >
      <div class="alarm-widget">
        <!-- Summary header -->
        <.alarm_summary
          critical={@counts.critical}
          warning={@counts.warning}
          info={@counts.info}
          on_acknowledge_all={
            @on_acknowledge &&
              JS.push("acknowledge_alarms", value: %{alarm_ids: Enum.map(@filtered_alarms, & &1.id)})
          }
        />
        
    <!-- Alarm list -->
        <div class="alarm-list mt-2">
          <.alarm_item
            :for={alarm <- @filtered_alarms}
            alarm={alarm_to_map(alarm)}
            acknowledged={alarm.acknowledged_at != nil || MapSet.member?(@acknowledged_ids, alarm.id)}
            on_acknowledge={@on_acknowledge && JS.push("acknowledge_alarm", value: %{id: alarm.id})}
            on_shelve={@on_shelve && JS.push("open_shelve_modal", value: %{id: alarm.id})}
            on_clear={@on_clear && JS.push("clear_alarm", value: %{id: alarm.id})}
          />

          <.widget_empty
            :if={@filtered_alarms == []}
            icon="hero-check-circle"
            message="No alarms"
          />
        </div>
        
    <!-- Filter indicator -->
        <div :if={@has_filters} class="alarm-filter-info">
          <div class="text-xs text-base-content/40 flex items-center gap-2">
            <.icon name="hero-funnel" class="w-3 h-3" />
            <span>Filtered view</span>
          </div>
        </div>
      </div>
    </.widget>
    """
  end

  defp filter_by_severity(alarms, []), do: alarms

  defp filter_by_severity(alarms, severities) do
    severity_atoms = Enum.map(severities, &String.to_existing_atom/1)
    Enum.filter(alarms, &(&1.severity in severity_atoms))
  end

  defp filter_by_target(alarms, nil), do: alarms

  defp filter_by_target(alarms, target_filter) do
    Enum.filter(alarms, fn alarm ->
      alarm.target_id == target_filter || alarm.target_name == target_filter
    end)
  end

  defp filter_cleared(alarms, true), do: alarms
  defp filter_cleared(alarms, _), do: Enum.reject(alarms, &(&1.status == :cleared))

  defp calculate_counts(alarms) do
    Enum.reduce(alarms, %{critical: 0, warning: 0, info: 0}, fn alarm, acc ->
      case alarm.severity do
        :critical -> Map.update!(acc, :critical, &(&1 + 1))
        :warning -> Map.update!(acc, :warning, &(&1 + 1))
        _ -> Map.update!(acc, :info, &(&1 + 1))
      end
    end)
  end

  defp alarm_to_map(alarm) do
    %{
      id: alarm.id,
      severity: alarm.severity,
      message: alarm.message,
      target_name: Map.get(alarm, :target_name) || Map.get(alarm, :source_id),
      triggered_at: alarm.triggered_at,
      name: Map.get(alarm, :alarm_type)
    }
  end

  defp widget_config_value(widget_id, widget_type, config) do
    %{widget_id: widget_id, widget_type: widget_type, config: config}
  end
end

File: legacy/cadence_legacy/assets/js/widgets/alarm_widget.js
/**
 * Alarm Summary Widget
 *
 * Filtered alarm view with inline actions.
 * Features:
 * - Configurable severity filter (critical/warning/info)
 * - Configurable source filter (by target, subsystem)
 * - Inline actions: Acknowledge, Shelve, Clear
 * - Visual severity indication with animated critical alarms
 * - Auto-scroll to new alarms
 */

import { BaseWidget, registerWidget } from "../dashboard/widget_factory"

/**
 * AlarmWidget - Alarm summary with inline actions.
 */
export class AlarmWidget extends BaseWidget {
  static displayName = "Alarm Summary"
  static description = "Filtered alarm list with inline actions"
  static icon = "bell"

  constructor(options) {
    super(options)
    this.type = "alarm"
    this.alarms = []
    this.acknowledgedIds = new Set()
  }

  init() {
    this.render()
    this.initialized = true
  }

  renderContent() {
    const {
      severity_filter = [], // empty = all, or ['critical', 'warning', 'info']
      target_filter = null,
      subsystem_filter = null,
      max_alarms = 20,
      show_cleared = false,
      auto_scroll = true
    } = this.config

    // Filter alarms based on configuration
    let filteredAlarms = [...this.alarms]

    if (severity_filter.length > 0) {
      filteredAlarms = filteredAlarms.filter(a => severity_filter.includes(a.severity))
    }

    if (target_filter) {
      filteredAlarms = filteredAlarms.filter(a => a.target_id === target_filter || a.target_name === target_filter)
    }

    if (subsystem_filter) {
      filteredAlarms = filteredAlarms.filter(a => a.subsystem === subsystem_filter)
    }

    if (!show_cleared) {
      filteredAlarms = filteredAlarms.filter(a => a.status !== 'cleared')
    }

    // Limit number of alarms shown
    filteredAlarms = filteredAlarms.slice(0, max_alarms)

    // Count by severity
    const counts = {
      critical: this.alarms.filter(a => a.severity === 'critical' && a.status !== 'cleared').length,
      warning: this.alarms.filter(a => a.severity === 'warning' && a.status !== 'cleared').length,
      info: this.alarms.filter(a => (a.severity === 'info' || !a.severity) && a.status !== 'cleared').length
    }

    return `
      <div class="alarm-widget flex flex-col h-full">
        <!-- Summary header -->
        <div class="alarm-summary flex items-center justify-between mb-2 pb-2 border-b border-base-300">
          <div class="flex items-center gap-3">
            ${counts.critical > 0 ? `
              <div class="flex items-center gap-1 ${counts.critical > 0 ? 'status-critical-pulse' : ''}">
                <span class="w-2 h-2 rounded-full bg-error"></span>
                <span class="text-xs font-bold text-error">${counts.critical}</span>
              </div>
            ` : ''}
            ${counts.warning > 0 ? `
              <div class="flex items-center gap-1">
                <span class="w-2 h-2 rounded-full bg-warning"></span>
                <span class="text-xs font-bold text-warning">${counts.warning}</span>
              </div>
            ` : ''}
            ${counts.info > 0 ? `
              <div class="flex items-center gap-1">
                <span class="w-2 h-2 rounded-full bg-info"></span>
                <span class="text-xs font-bold text-info">${counts.info}</span>
              </div>
            ` : ''}
            ${counts.critical === 0 && counts.warning === 0 && counts.info === 0 ? `
              <div class="flex items-center gap-1">
                <span class="w-2 h-2 rounded-full bg-success"></span>
                <span class="text-xs text-success">All clear</span>
              </div>
            ` : ''}
          </div>
          <div class="flex items-center gap-1">
            <button class="btn btn-ghost btn-xs" id="alarm-ack-all-${this.id}" title="Acknowledge All">
              <svg class="w-3 h-3" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M5 13l4 4L19 7"/>
              </svg>
            </button>
          </div>
        </div>

        <!-- Alarm list -->
        <div class="alarm-list flex-1 overflow-y-auto space-y-1" id="alarm-list-${this.id}">
          ${filteredAlarms.length > 0 ? filteredAlarms.map(alarm => this._renderAlarmItem(alarm)).join('') : `
            <div class="text-center py-8 text-base-content/40">
              <svg class="w-8 h-8 mx-auto mb-2 opacity-50" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"/>
              </svg>
              <p class="text-xs">No alarms</p>
            </div>
          `}
        </div>

        <!-- Filter indicator -->
        ${(severity_filter.length > 0 || target_filter || subsystem_filter) ? `
          <div class="alarm-filter-info mt-2 pt-2 border-t border-base-300">
            <div class="text-xs text-base-content/40 flex items-center gap-2">
              <svg class="w-3 h-3" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M3 4a1 1 0 011-1h16a1 1 0 011 1v2.586a1 1 0 01-.293.707l-6.414 6.414a1 1 0 00-.293.707V17l-4 4v-6.586a1 1 0 00-.293-.707L3.293 7.293A1 1 0 013 6.586V4z"/>
              </svg>
              <span>Filtered view</span>
            </div>
          </div>
        ` : ''}
      </div>
    `
  }

  _renderAlarmItem(alarm) {
    const severityClass = this._getSeverityClass(alarm.severity)
    const isAcknowledged = this.acknowledgedIds.has(alarm.id) || alarm.acknowledged
    const isNew = alarm.isNew // Set by updateAlarms when a new alarm comes in

    return `
      <div
        class="alarm-item ${severityClass} ${isNew ? 'alarm-new' : ''} ${isAcknowledged ? 'opacity-60' : ''}"
        data-alarm-id="${alarm.id}"
      >
        <div class="alarm-item-header flex items-start justify-between gap-2">
          <div class="flex items-center gap-2 min-w-0 flex-1">
            <span class="alarm-severity-dot w-2 h-2 rounded-full flex-shrink-0 ${this._getSeverityDotClass(alarm.severity)}"></span>
            <span class="alarm-source text-xs font-medium truncate">${alarm.target_name || 'Unknown'}</span>
          </div>
          <span class="alarm-time text-xs text-base-content/40 flex-shrink-0">${this._formatTime(alarm.triggered_at)}</span>
        </div>

        <div class="alarm-message text-sm mt-1 ${alarm.severity === 'critical' ? 'font-medium' : ''}">
          ${alarm.message || alarm.name || 'Alarm'}
        </div>

        <div class="alarm-actions flex items-center gap-1 mt-2">
          ${!isAcknowledged ? `
            <button class="alarm-btn alarm-btn-ack" data-action="acknowledge" title="Acknowledge">
              ACK
            </button>
          ` : `
            <span class="text-xs text-base-content/40">Acknowledged</span>
          `}
          <button class="alarm-btn" data-action="shelve" title="Shelve">
            SHELVE
          </button>
          <button class="alarm-btn alarm-btn-clear" data-action="clear" title="Clear">
            CLEAR
          </button>
          <button class="alarm-btn ml-auto" data-action="details" title="View Details">
            <svg class="w-3 h-3" fill="none" stroke="currentColor" viewBox="0 0 24 24">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M13 16h-1v-4h-1m1-4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"/>
            </svg>
          </button>
        </div>
      </div>
    `
  }

  _getSeverityClass(severity) {
    switch (severity) {
      case 'critical':
        return 'alarm-item-critical'
      case 'warning':
        return 'alarm-item-warning'
      case 'info':
      default:
        return 'alarm-item-info'
    }
  }

  _getSeverityDotClass(severity) {
    switch (severity) {
      case 'critical':
        return 'bg-error'
      case 'warning':
        return 'bg-warning'
      case 'info':
      default:
        return 'bg-info'
    }
  }

  _formatTime(timestamp) {
    if (!timestamp) return '--:--'
    const date = new Date(timestamp)
    return date.toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false
    })
  }

  render() {
    super.render()
    this._bindActions()
  }

  _bindActions() {
    // Acknowledge all button
    this.container.querySelector(`#alarm-ack-all-${this.id}`)?.addEventListener('click', (e) => {
      e.stopPropagation()
      this._dispatchAction('acknowledge_all', null)
    })

    // Individual alarm actions
    this.container.querySelectorAll('.alarm-item').forEach(item => {
      const alarmId = item.dataset.alarmId

      item.querySelectorAll('.alarm-btn').forEach(btn => {
        btn.addEventListener('click', (e) => {
          e.stopPropagation()
          const action = btn.dataset.action
          this._dispatchAction(action, alarmId)

          // Optimistic UI update for acknowledge
          if (action === 'acknowledge') {
            this.acknowledgedIds.add(alarmId)
            this.render()
          }
        })
      })
    })
  }

  _dispatchAction(action, alarmId) {
    const event = new CustomEvent("widget:alarm_action", {
      detail: {
        widgetId: this.id,
        alarmId: alarmId,
        action: action
      },
      bubbles: true
    })
    this.container.dispatchEvent(event)
  }

  // Called by external code to update alarms
  updateAlarms(alarms, options = {}) {
    const { markNewAlarms = true } = options

    if (markNewAlarms) {
      // Mark new alarms that weren't in the previous list
      const previousIds = new Set(this.alarms.map(a => a.id))
      alarms = alarms.map(a => ({
        ...a,
        isNew: !previousIds.has(a.id)
      }))
    }

    this.alarms = alarms
    this.render()

    // Auto-scroll to top if configured and there are new alarms
    if (this.config.auto_scroll && alarms.some(a => a.isNew)) {
      const list = this.container.querySelector(`#alarm-list-${this.id}`)
      if (list) {
        list.scrollTop = 0
      }
    }

    // Clear "new" markers after a delay
    setTimeout(() => {
      this.alarms = this.alarms.map(a => ({ ...a, isNew: false }))
      // Don't re-render, just clear the flag for next update
    }, 3000)
  }

  // Add a single alarm
  addAlarm(alarm) {
    this.alarms = [{ ...alarm, isNew: true }, ...this.alarms]
    this.render()

    if (this.config.auto_scroll) {
      const list = this.container.querySelector(`#alarm-list-${this.id}`)
      if (list) {
        list.scrollTop = 0
      }
    }
  }

  // Remove an alarm
  removeAlarm(alarmId) {
    this.alarms = this.alarms.filter(a => a.id !== alarmId)
    this.acknowledgedIds.delete(alarmId)
    this.render()
  }

  openConfig() {
    const event = new CustomEvent("widget:configure", {
      detail: {
        widgetId: this.id,
        widgetType: this.type,
        config: this.getConfig()
      },
      bubbles: true
    })
    this.container.dispatchEvent(event)
  }
}

// Register the widget
registerWidget("alarm", AlarmWidget)

export default AlarmWidget
